change/@itwin-tree-widget-react-d9ea273e-f06f-4880-bb8b-a70991486fd9.json

```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Improve performance of visibility status calculation and visibility changes. Also reduce main thread blocking to avoid UI freezing.",
```

Comment:
Agree, it would be great if the PR description mentioned what improvement to expect. I understand it's not trivial to get the numbers for this kind of problem, but something approximate could be retrieved from the devtools?.. E.g., take one specific test case and provide something like this:

|                                                                                                         | Before the changes | After the changes |
| ------------------------------------------------------------------------------------------------------- | ------------------ | ----------------- |
| Time it takes to fully determine nodes' visibility for initial view, filtered by X target elements      | ?                  | ?                 |
| Time it takes to fully handle root node visibility change, with hierarchy filtered by X target elements | ?                  | ?                 |
| Number of main thread blocks above 1s                                                                   | ?                  | ?                 |
| Number of main thread blocks above 500ms                                                                | ?                  | ?                 |
| Number of main thread blocks above 100ms                                                                | ?                  | ?                 |
| Number of main thread blocks above 50ms                                                                 | ?                  | ?                 |

Reply:
Updated description